### ISSUE TITLE
How will data platform provider data be represented by the API?

### ISSUE THREAD

**Issue by reefdog** [PhilanthropyDataCommons/service#156]
<details>
<summary>Prerequisite assumption (click to view).</summary>

_I'm hiding this behind a collapse because it's not **directly** relevant to the issue, but is an important prerequisite assumption._

Applicants seeded from different external sources will _not_ be merged into a single entity in the PDC or represented as a single entity by the API, at least during the pilot. That is, an applicant from External Source A who could be reliably identified as the same entity as an applicant from External Source B will still exist as two discrete applicants in the PDC and via the API, at least initially. This is fine and expected.

----

</details>

As an API consumer (and developer of the [Data Viewer](https://github.com/PhilanthropyDataCommons/data-viewer/)), I plan to fetch applicants and display their attributes to my users. I will also want to supplement this with additional info from "data platform providers" such as Charity Navigator, and distinguish between data that came from an application and data that came from a data platform provider.

Technically, how will I do this?

Let's say two funders each send the PDC an applicant with the EIN `99-9999999`. That creates two `applicant` records and all their associated fields. A call to the `/applicants` endpoint returns these two applicants, which the Data Viewer will display as separate records (until we figure out if/how to associate them, as noted in the prerequisite).

Now let's say the PDC scrapes a data platform provider for additional info about the organization attached to EIN `99-9999999`. How is this data made available to the API?

I can see either of these being the case:

1. The PDC service virtually stitches the data platform provider-sourced fields into all matching applicants, such that `/applicants` still only returns two applicants (those that came from the funders), but now with additional attributes provided by the data platform provider and noted as such.
   <details>
   <summary>View JSON pseudo-model of this case.</summary>

   ```json
   [
     {
       "id": "1c197e93-51ea-473c-a427-19d1e68db0ad",
       "ein": "99-9999999",
       "name": "Do Good Things",
       "externalSource": "Jane Doe Foundation",
       "dataPlatformProviderData": [
         {
           "source": "Charity Navigator",
           "organizationName": "Do Good Things, LLC"
         }
       ]
     },
     {
       "id": "e6f551c4-706e-4590-9e08-71331891bc49",
       "ein": "99-9999999",
       "name": "Do Good Things",
       "externalSource": "The Amazing Foundation",
       "dataPlatformProviderData": [
         {
           "source": "Charity Navigator",
           "organizationName": "Do Good Things, LLC"
         }
       ]
     },
   ]
   ```

   </details>
2. The PDC service provides the data platform provider-sourced fields as a third applicant on the `/applicants` endpoint, and it's up to me (the API consumer) to decide whether to display them differently based on source.
   <details>
   <summary>View JSON pseudo-model of this case.</summary>

   ```json
   [
     {
       "id": "1c197e93-51ea-473c-a427-19d1e68db0ad",
       "ein": "99-9999999",
       "name": "Do Good Things",
       "externalSource": "Jane Doe Foundation",
     },
     {
       "id": "e6f551c4-706e-4590-9e08-71331891bc49",
       "ein": "99-9999999",
       "name": "Do Good Things",
       "externalSource": "The Amazing Foundation",
     },
     {
       "id": "18f1502e-b7fb-48e1-9382-cdfd0d733347",
       "ein": "99-9999999",
       "name": "Do Good Things, LLC",
       "externalSource": "Charity Navigator",
     },
   ]
   ```

   </details>

_Side note: I'm aware that the current API doesn't actually return applicant data this way, and that I actually might be composing an applicant from something like `ExternalFieldValue`s. So consider the above a JSON representation of a relational concept._

**Assuming I've framed this correctly so far**: What is the preferred model?

In ①, the job of identifying data platform providers is on the PDC service, and the API consumer will just use the data structure to know which is which. In ②, the job is on the API consumer to treat data sources differently; however, the API consumer will still need to identify data platform providers among the data sources, either by looking for a specific attribute on the external source (which we would need to add to the data model) or by hard-coding a list of known data platform providers (eww).

Please note that I've framed this issue around "how will it look to the API" because I'm greedily thinking about my own needs as an API consumer, but it's also a discussion about internal PDC design.

**Comment by bickelj** [comment]
I think either way the data model should have a field to help distinguish attributes of applicants that come from funders versus attributes that come from anywhere else. And that change would come first regardless of the direction taken above. As a back-end engineer I would of course rather punt the stitching together to the front-end and lean toward option ② but option ① makes sense as well. There might be less selfish reasons to prefer the latter model as well, but I think it's probably going to be personal preference or performance reasons.

<details>
<summary>Example</summary>

For example the following JSON
```json
[
  {
    "id": "1c197e93-51ea-473c-a427-19d1e68db0ad",
    "ein": "99-9999999",
    "name": "Do Good Things",
    "externalSource": "Jane Doe Foundation",
  },
  {
    "id": "e6f551c4-706e-4590-9e08-71331891bc49",
    "ein": "99-9999999",
    "name": "Do Good Things",
    "externalSource": "The Amazing Foundation",
  },
  {
    "id": "18f1502e-b7fb-48e1-9382-cdfd0d733347",
    "ein": "99-9999999",
    "name": "Do Good Things, LLC",
    "externalSource": "Charity Navigator",
  },
]
```
becomes this JSON

```json
[
  {
    "id": "1c197e93-51ea-473c-a427-19d1e68db0ad",
    "ein": "99-9999999",
    "name": "Do Good Things",
    "externalSource": "Jane Doe Foundation",
  },
  {
    "id": "e6f551c4-706e-4590-9e08-71331891bc49",
    "ein": "99-9999999",
    "name": "Do Good Things",
    "externalSource": "The Amazing Foundation",
  },
  {
    "id": "18f1502e-b7fb-48e1-9382-cdfd0d733347",
    "ein": "99-9999999",
    "name": "Do Good Things, LLC",
    "externalSource": "Charity Navigator",
  },
  {
    "id": "12345",
    "ein": "99-9999999",
    "name": "Do Good Things, LLC",
    "externalSource": "Some other external source",
  },
]
```
</details>

<details>
<summary>P.S.</summary>
Thanks for the example of using `details` in a post. It is a really helpful tool.
</details>

**Comment by reefdog** [comment]
> I think either way the data model should have a field to help distinguish attributes of applicants that come from funders versus attributes that come from anywhere else.

I assume so as well.

> As a back-end engineer I would of course rather punt the stitching together to the front-end and lean toward option ② but option ① makes sense as well. There might be less selfish reasons to prefer the latter model as well, but I think it's probably going to be personal preference or performance reasons.

I agree. The primary determinant should be which option best reflects the design of the PDC and the conceptual data model. (E.g., should the PDC be modeling infomediary data as peer or supplementary to the data that comes in via funders?)

I do expect to rapidly wish for API endpoints more targeted to my use case, for ergonomic and performance reasons. But I'm fine starting with letting the PDC be its best self regardless of the data viewer's needs.

**Comment by reefdog** [comment]
Turns out this whole issue was framed around a misunderstanding of the [ERD](https://github.com/PhilanthropyDataCommons/service/blob/main/docs/ENTITY_RELATIONSHIP_DIAGRAM.md).

Data from data platform providers will be captured as `ExternalFieldValue`s, with the platform itself identified as being an `ExternalSource`. To the extent that we identify "first-party" providers of proposal data (funders, GMSes, etc.), those will be identified as something other than an `ExternalSource`.

So the short answer: `ExternalSource`s are data platform providers, and `ExternalFieldValue`s are the attributes about an `Applicant` that we've scraped.
